Feat/issue 303: de-duplicate plot spikes code

AdityaBITMESRA opened the pull request:
I have created a pull request with the necessary instructions.

sanjayankur31 commented:
@AdityaBITMESRA : could you rebase your branch against development and update the PR to use development as the base branch too please?

AdityaBITMESRA commented:
I have changed the base branch from master to development

sanjayankur31 commented:
> I have changed the base branch from master to development

hrm, now we have conflicts again---did you create your feature branch from master or development? It should be off development and then we shouldn't have any conflicts :thinking:

AdityaBITMESRA commented:
yes sir I think I created it against master branch...I will resolve this

sanjayankur31 commented:
No worries. Maybe a simple rebase will work:

```
git checkout feat/issue-303
git rebase development
```

that'll take your branch and "rebase" it on to development from master. It may not work given how master and development have now diverged, but worth a try. if your commits are small enough, you can even create a new feat/issue-303-1 branch from development and cherri-pick them?

More on both of these here:

- https://git-scm.com/book/en/v2/Git-Branching-Rebasing
- https://git-scm.com/docs/git-cherry-pick

AdityaBITMESRA commented:
I think the conflicts are resolved. Can you verify sir @sanjayankur31

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -37,9 +36,6 @@
 
 
 def process_args():
-    """
```
Please keep the docstrings (and improve them?)

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -106,18 +102,7 @@ def process_args():
-
-
-def read_sonata_spikes_hdf5_file(file_name: str):
-    """Read a sonata HDF5 file with spike data
```
Please keep the docstrings. Every function should have one. This is required for the API docs:

https://pyneuroml.readthedocs.io/en/development/

sanjayankur31 reviewed (changes requested):
No conflicts now. I'll do a full review by the end of the week. I've dropped a few comments to begin with---please keep the docstrings and ensure they are up to date. Please also add type hints to function arguments etc. All of this helps with linting and doc generation.

https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html

AdityaBITMESRA commented:
Sure sir, sorry for the late reply. I will straightaway work on the instructions

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -108,7 +108,16 @@ def process_args() -> argparse.Namespace:
+        Read spike times from a SONATA format HDF5 file.
+
+    Args:
+        file_name (str): The name of the HDF5 file.
```
ah, good, but please follow the same format as the other docstrings

AdityaBITMESRA replied:
Okay sir

AdityaBITMESRA commented:
Hi sir @sanjayankur31 , I have updated the code with the necessary type hints and docstrings. Can you please review it?

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -187,12 +185,54 @@ def read_sonata_spikes_hdf5_file(file_name: str):
-        "Generating spiketime plot for %s; format: %s; plotting: %s; save to: %s"
-        % (a.spiketime_files, a.format, a.show_plots_already, a.save_spike_plot_to)
-    )
+from typing import Dict, List, Optional, Union
```
move these to the top of the file (imports should be at the top usually)

Hint: you can use something like [isort](https://pypi.org/project/isort/) to do this for you. It integrates with lots of IDEs, so check to see how to use it for your setup.gg

AdityaBITMESRA commented:
Okay sir

AdityaBITMESRA commented:
Hello sir @sanjayankur31, I'm working on the test_plot_spikes_from_sonata_file test case in test_plot_spikes.py. It assumes a function called write_spike_data_to_hdf5 exists to write spike data to HDF5 in SONATA format, but I can't find its implementation. Could you clarify if this function exists and where it's defined? If not, how should I proceed with this test case?

sanjayankur31 commented:
> Hello sir @sanjayankur31, I'm working on the test_plot_spikes_from_sonata_file test case in test_plot_spikes.py. It assumes a function called write_spike_data_to_hdf5 exists to write spike data to HDF5 in SONATA format, but I can't find its implementation. Could you clarify if this function exists and where it's defined? If not, how should I proceed with this test case?

I don't think we've implemented a function to write spike files to HDF5 format yet so `write_spike_data_to_hdf5` doesn't exist. Maybe just add a test function but mark it to be skipped for the moment. I'll see if I can find a SONATA spikes file that we can use to test here.

AdityaBITMESRA commented:
I have added test_PlotSpikes.py file @sanjayankur31

sanjayankur31 commented:
@all-contributors please add @AdityaBITMESRA for bug, ideas

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -36,9 +36,12 @@
+    Parse command line arguments.
+
+    Returns:
+        argparse.Namespace: The parsed arguments.
```
cosmetic: please use the `:returns: ..` style

sanjayankur31 reviewed (commented):
This is looking very good.

I was looking at how we implemented the time series plotting. What we did was to have three functions:

- `plot_time_series`: this takes data and plots it
- `plot_time_series_from_data_files`:  this reads the files, puts the data into the format required by `plot_time_series` and calls it
- `plot_time_series_from_lems_file`: this one is special---it reads a LEMS simulation file, figures out the data files from there, calls `plot_time_series_from_data_files` etc.

So I was wondering if we should follow the same system for the spike plotting bit:

- `plot_spikes`: this takes data and plots it
- `plot_spikes_from_data_files`: this reads files, puts the data in to the format required by `plot_spikes` and calls it to do the plotting
- `plot_spikes_from_lems_file`: reads a LEMS simulation file to figure out the names of the data files etc, then calls `plot_spikes_from_data_files` to load them etc.

What do you think of this?

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -36,9 +36,12 @@
 
 
-def process_args():
+def process_args() -> argparse.Namespace:
```
I think we should make this `_process_args` to indicate that it's an internal method and not part of the public API?

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -381,13 +421,24 @@ def run(a=None, **kwargs):
         plt.close()
 
 
+def main(args: Optional[argparse.Namespace] = None) -> None:
```
Let's rename this to `_spike_plotter_main` for clarity?

AdityaBITMESRA commented:
Thankyou sir for all the clear instructions, I will work on this asap

sanjayankur31 commented:
Is this ready for another round of review @AdityaBITMESRA ?

AdityaBITMESRA commented:
yes sir @sanjayankur31 I have updated the code with the necessary instructions but it is failing the checks, so I am resolving them.

sanjayankur31 commented:
Sounds good. You should be able to run all the tests on your system locally using the `test.sh` script too. If only, unit tests, you can use `pytest` directly---and you can even tell pytest what specific test to run using the `-k` option. See `pytest -h` for usage instructions.

AdityaBITMESRA commented:
Hi sir @sanjayankur31 , all checks have passed so I request to review once again if time permits

sanjayankur31 commented:
Thanks, I've just returned from leave, so I'll take a look this week.

AdityaBITMESRA commented:
Sure sir...

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -381,13 +317,192 @@ def run(a=None, **kwargs):
+    :return: A tuple containing a list of spike data file paths and the format of the spike data files.
+    :rtype: Tuple[List[str], str]
+    """
+    # Code to read the LEMS file and extract the spike data file paths and format
```
@AdityaBITMESRA : this looks incomplete?

sanjayankur31 replied:
You should also be using the `pynml.load_sim_data_from_lems_file` function here to do the loading. Again, see `PlotTimeSeries.py`.

sanjayankur31 commented on tests/plot/test_PlotSpikes.py:
```diff
@@ -0,0 +1,109 @@
+        os.unlink("spike-plot-test.png")
+
+    def test_plot_spikes_from_files(self):
+        """Test the plot_spikes function with spike time files."""
```
This should be reading spike times from files, doesn't seem to be doing so?

sanjayankur31 replied:
@AdityaBITMESRA this hasn't been updated yet---it needs to read data from the file of spikes that you've created, not from the `spike_data` variable

sanjayankur31 commented:
I'm not sure it works properly. I ran the `test_plot_spikes_from_data` test and got this:

```
pytest -k test_plot_spikes_from_data
```

![image](https://github.com/NeuroML/pyNeuroML/assets/102575/171078f6-e4d5-48a6-8762-901ff875d801)


I don't see any spikes here.  Is it working on your system @AdityaBITMESRA ?

sanjayankur31 commented:
the `pynml-plotspikes` command isn't working either from the looks of it:

```
$ cat KC_step_test.KC_pop.spikes 
0       0.14395000000004157
0       0.17676000000001174
0       0.2092299999999822
0       0.24151999999995283
0       0.27365999999992363
0       0.3056799999998945
0       0.3375699999998655
0       0.3693299999998366
0       0.4009399999998079
0       0.4323799999997793
0       0.46361999999975084
0       0.4945999999997227
0       0.5252799999996949
0       0.5555899999996673
0       0.5854499999996401
(ins)(neuroml-311-dev)[asinha@thor  pyNeuroML(feat/issue-303 *%=)]$ pynml-plotspikes KC_step_test.KC_pop.spikes 
Traceback (most recent call last):
  File "/home/asinha/.local/share/virtualenvs/neuroml-311-dev/bin/pynml-plotspikes", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/asinha/Documents/02_Code/00_mine/NeuroML/software/pyNeuroML/pyneuroml/plot/PlotSpikes.py", line 484, in main
    if args.lemsFile:
       ^^^^^^^^^^^^^
AttributeError: 'Namespace' object has no attribute 'lemsFile'

```

Can you please look into these? Please do test out the various functions one by one to make sure they work. The tests only check that a png file is created, not that it is a sensible plot :)

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -381,13 +317,192 @@ def run(a=None, **kwargs):
+    :rtype: None
+    """
+    # Code to read the LEMS simulation file and get the spike data file paths and format
+    spike_data_files, spike_data_format = get_spike_data_files_from_lems(
```
this is a no-op at the moment

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -381,13 +317,192 @@ def run(a=None, **kwargs):
+    """
+    if args is None:
+        args = _process_args()
+    if args.lemsFile:
```
there is no `-lemsFile` argument. The idea is that we detect if the provided files are LEMS files or data files containing spikes. See how this is handled in `PlotTimeSeries.py`---just check that the files start with `LEMS_`, for example.

sanjayankur31 reviewed (changes requested):
It's getting there, but the functionality is not complete. It doesn't currently work properly, so you'll need to see what else needs to be implemented here.

AdityaBITMESRA commented:
Thankyou for looking into it sir @sanjayankur31. I will like to borrow some time so that I go through all the issues and correct them.

AdityaBITMESRA commented:
Going through all the issues you pointed out ,they seem to be functional and can be solved if I take reference from PlotTimeSeries file, right?

sanjayankur31 commented:
Yes, you should be able to refer to plot-time-series. :+1:

AdityaBITMESRA commented:
sir @sanjayankur31 I have updated the get_spike_data_files_from_lems function to add the necessary functionality and also removed the -lemsFile argument from the main function to align the logic with PlotTimeSeries function  for lems file.

AdityaBITMESRA commented:
I am really confused about not seeing any spikes after running the test_plot_spikes_from_data as pointed out earlier by you. I have tried to align with the logic of production plotspikes code but still not getting desired results. Is there anything which you would like to add?

sanjayankur31 commented:
I haven't had a chance to go over the code in detail yet. Maybe try with more neurons and more spikes---maybe there are too few to be able for us to see?

AdityaBITMESRA commented:
![Spike_times_from:_](https://github.com/NeuroML/pyNeuroML/assets/147182147/b85a891a-cb10-4e54-97dc-9a62897b94fa)
sir @sanjayankur31 I increased the number of neuron and spike data in test file and this was the plot. What do you think about it?

sanjayankur31 commented:
Maybe check your test data to ensure the cell ids are unique, and that each cell only belongs to one population

AdityaBITMESRA commented:
![Spike_times_from:_](https://github.com/NeuroML/pyNeuroML/assets/147182147/f46ddeaf-93b9-496c-9b40-0ecbcff15c50)
I made the necessary changes . Please take a look sir @sanjayankur31

AdityaBITMESRA commented:
![Screenshot from 2024-04-17 21-31-05](https://github.com/NeuroML/pyNeuroML/assets/147182147/3df289ab-563e-49cb-8ff7-bf3c5b2c6bd5)
I made the necessary changes. Please take a look sir @sanjayankur31

sanjayankur31 commented:
That looks correct now. so we're getting there. Please see why the tests are failing---once they pass I'll do another round of review.

AdityaBITMESRA commented:
sure sir

sanjayankur31 commented:
The tests are failing with:

```
FAILED tests/plot/test_PlotSpikes.py::TestPlotSpikes::test_plot_spikes_from_files - ValueError: Image size of 876x620685 pixels is too large. It must be less than 2^16 in each direction
```

too much data in the test perhaps, or maybe see if you can reduce the DPI etc to reduce the size of the image. 

Does it work when you run it locally?

AdityaBITMESRA commented:
yes sir the plot was created but the test session was unusually very long and I ignored that it failed(1 failed,1 passed) because of the image size. I am resolving this

AdityaBITMESRA commented:
Hello sir, @sanjayankur31 while going through the original PlotSpikes.py file which is in production, I found out that it had explicit error handling mechanisms using try-except blocks, which are missing in the refactored code.Should I work on reintroducing these mechanisms to ensure the code's reliability

AdityaBITMESRA commented:
Hello sir @sanjayankur31, I have been trying to write the test function for test_plot_spikes_from_lems_files but I am encountering errors. I have taken the test_plot_time_series_from_lems_file as reference for writing. So am I right with my approach?

sanjayankur31 commented:
> Hello sir, @sanjayankur31 while going through the original PlotSpikes.py file which is in production, I found out that it had explicit error handling mechanisms using try-except blocks, which are missing in the refactored code.Should I work on reintroducing these mechanisms to ensure the code's reliability

Yes---let's make it as reliable as we can



> Hello sir @sanjayankur31, I have been trying to write the test function for test_plot_spikes_from_lems_files but I am encountering errors. I have taken the test_plot_time_series_from_lems_file as reference for writing. So am I right with my approach?

That sounds correct. What error are you seeing? Maybe we can help?

AdityaBITMESRA commented:
Sir I got an index error (IndexError: list index out of range) and it is an error of the **load_sim_data_from_lems_file** function.
Also can you tell me why we have not used plot_time_series_from_lems_file(plot_time_series_from_data_files was used) function in the lems test case

sanjayankur31 commented:
> Sir I got an index error (IndexError: list index out of range) and it is an error of the **load_sim_data_from_lems_file** function. 

Right, can you paste your code and the exact error? You need to make sure that the lems file you're using does in fact have some files that are recording spikes listed in them using the `EventOutputFile` bit:

https://github.com/sanjayankur31/262670/blob/neuroml-ankur/NeuroML2/LEMS_KC_step_test.xml#L39


> Also can you tell me why we have not used plot_time_series_from_lems_file(plot_time_series_from_data_files was used) function in the lems test case of **test_PlotTimeSeries** function

This looks like a mistake. it should be using `lems_file` there instead of `trace_file`. Thanks for pointing this out, I'll fix it now.

AdityaBITMESRA commented:
import os
import tempfile
import numpy as np


def test_plot_spikes_from_lems_file(self):
    """Test plot_spikes_from_lems_file function"""
    spike_file = tempfile.NamedTemporaryFile(mode="w", delete=False, dir=".")
    for i in range(0, 1000):
        print(f"{i/1000}\t{np.random.default_rng().random()}\t{np.random.default_rng().random()}\t{np.random.default_rng().random()}", file=spike_file)
    spike_file.flush()
    spike_file_path = spike_file.name
    spike_file.close()

    lems_contents = f"""
<Lems>
    
    <!-- Specify which component to run -->
    <Target component="example_izhikevich2007network_sim"/>
    <!-- Include core NeuroML2 ComponentType definitions -->
    <Include file="Cells.xml"/>
    <Include file="Networks.xml"/>
    <Include file="Simulation.xml"/>
    <Include file="izhikevich2007_network.nml"/>
    <Simulation id="example_izhikevich2007network_sim" length="1000ms" step="0.1ms" target="IzNet" seed="123">
        <!-- Note seed: ensures same random numbers used every run -->
        <OutputFile id="output0" fileName={xml.sax.saxutils.quoteattr(spike_file_path)}>
            <OutputColumn id="IzhPop0[0]" quantity="IzhPop0[0]/v"/>
        </OutputFile>
        <Recording>
            <EventSelection select="EVENT_OUT">
                <Population name="Population1">
                    <Event time="1.0" id="1"/>
                    <Event time="2.0" id="2"/>
                    <Event time="3.0" id="3"/>
                </Population>
                <Population name="Population2">
                    <Event time="2.5" id="4"/>
                    <Event time="3.5" id="5"/>
                </Population>
            </EventSelection>
        </Recording>
    </Simulation>
</Lems>
"""

    lems_file = tempfile.NamedTemporaryFile(mode="w", delete=False, dir=".")
    lems_file.write(lems_contents)
    lems_file.flush()
    lems_file_path = lems_file.name
    lems_file.close()

    pyplts.plot_spikes_from_lems_file(
        lems_file_name=lems_file_path,
        show_plots_already=False,
        save_spike_plot_to="spike-plot-from-lems-test.png",
    )
    print("plot_spikes_from_lems_file returned.")
    assert os.path.isfile("spike-plot-from-lems-test.png")

    os.unlink("spike-plot-from-lems-test.png")
    os.unlink(lems_file_path)
    os.unlink(spike_file_path)
    
    I was using this lems file which seems to be wrong

sanjayankur31 commented:
Yeh, check that the lems file is set up correctly. That should do it.

I fixed the test too: https://github.com/NeuroML/pyNeuroML/commit/a2066bcc94619c434d2c2aeb791e59d544a7f383

Do remember keep pulling the `development` branch into your branch here so that you have the latest code.

AdityaBITMESRA commented:
Hello sir @sanjayankur31 I was able to successfully run the test_plot_spikes_from_lems_file function and this was the plot created. Is this accurate?
 <EventOutputFile id="Spikes_file__IzhPop0" fileName="{spike_file_name}" format="ID_TIME">
            <EventSelection id="0" select="IzhPop0[0]" eventPort="spike">
                <SpikeData>
                    <Spike time="10" id="0"/>
                    <Spike time="20" id="0"/>
                    <Spike time="30" id="0"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="1" select="IzhPop0[1]" eventPort="spike">
                <SpikeData>
                    <Spike time="15" id="1"/>
                    <Spike time="25" id="1"/>
                    <Spike time="35" id="1"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="2" select="IzhPop0[2]" eventPort="spike">
                <SpikeData>
                    <Spike time="12" id="2"/>
                    <Spike time="22" id="2"/>
                    <Spike time="32" id="2"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="3" select="IzhPop0[3]" eventPort="spike">
                <SpikeData>
                    <Spike time="18" id="3"/>
                    <Spike time="28" id="3"/>
                    <Spike time="38" id="3"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="4" select="IzhPop0[4]" eventPort="spike">
                <SpikeData>
                    <Spike time="20" id="4"/>
                    <Spike time="30" id="4"/>
                    <Spike time="40" id="4"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="5" select="IzhPop0[5]" eventPort="spike">
                <SpikeData>
                    <Spike time="22" id="5"/>
                    <Spike time="32" id="5"/>
                    <Spike time="42" id="5"/>
                </SpikeData>
            </EventSelection>
        </EventOutputFile>
        
        This was the data of EventOutputFile bit



 ![Screenshot from 2024-04-30 13-25-39](https://github.com/NeuroML/pyNeuroML/assets/147182147/c2740090-485a-4348-aba4-6535d51bb21e)

AdityaBITMESRA commented:
<Lems>

    <!--

        This LEMS file has been automatically generated using PyNeuroML v1.1.13 (libNeuroML v0.5.8)

     -->

    <!-- Specify which component to run -->
    <Target component="example_izhikevich2007network_sim"/>

    <!-- Include core NeuroML2 ComponentType definitions -->
    <Include file="Cells.xml"/>
    <Include file="Networks.xml"/>
    <Include file="Simulation.xml"/>

    <Include file="izhikevich2007_network.nml"/>

    <Simulation id="example_izhikevich2007network_sim" length="1000ms" step="0.1ms" target="IzNet" seed="123">  <!-- Note seed: ensures same random numbers used every run -->
        <EventOutputFile id="Spikes_file__IzhPop0" fileName="{spike_file_name}" format="ID_TIME">
            <EventSelection id="0" select="IzhPop0[0]" eventPort="spike">
                <SpikeData>
                    <Spike time="10" id="0"/>
                    <Spike time="20" id="0"/>
                    <Spike time="30" id="0"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="1" select="IzhPop0[1]" eventPort="spike">
                <SpikeData>
                    <Spike time="15" id="1"/>
                    <Spike time="25" id="1"/>
                    <Spike time="35" id="1"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="2" select="IzhPop0[2]" eventPort="spike">
                <SpikeData>
                    <Spike time="12" id="2"/>
                    <Spike time="22" id="2"/>
                    <Spike time="32" id="2"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="3" select="IzhPop0[3]" eventPort="spike">
                <SpikeData>
                    <Spike time="18" id="3"/>
                    <Spike time="28" id="3"/>
                    <Spike time="38" id="3"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="4" select="IzhPop0[4]" eventPort="spike">
                <SpikeData>
                    <Spike time="20" id="4"/>
                    <Spike time="30" id="4"/>
                    <Spike time="40" id="4"/>
                </SpikeData>
            </EventSelection>
            <EventSelection id="5" select="IzhPop0[5]" eventPort="spike">
                <SpikeData>
                    <Spike time="22" id="5"/>
                    <Spike time="32" id="5"/>
                    <Spike time="42" id="5"/>
                </SpikeData>
            </EventSelection>
        </EventOutputFile>

    </Simulation>

</Lems>
This was my lems file. I used the lems file from the plottimeseries and enhanced it with EventOutputFile bit

sanjayankur31 commented:
Maybe reduce the size of the markers so we can see if they're being plotted at the right time?

Next---A LEMS file can include the spike data, but it is far more common for the LEMS file to point to an EventOutputFile that stores the spikes. So that'll be a proper test. See this example:

https://docs.neuroml.org/Userdocs/IzhikevichNetworkExample.html#simulating-the-model

(If you haven't been through these examples, it'll be good to do so. You do need to understand a little bit of how NeuroML/LEMS works)

AdityaBITMESRA commented:
Hello sir @sanjayankur31 I have made the changes and this is the plot 
![Screenshot from 2024-05-01 11-15-20](https://github.com/NeuroML/pyNeuroML/assets/147182147/ba1034b9-1a79-41c2-8dbb-d8bdf0b1ec28)

AdityaBITMESRA commented:
<Lems>

    <!--

        This LEMS file has been automatically generated using PyNeuroML v1.1.13 (libNeuroML v0.5.8)

     -->

    <!-- Specify which component to run -->
    <Target component="example_izhikevich2007network_sim"/>

    <!-- Include core NeuroML2 ComponentType definitions -->
    <Include file="Cells.xml"/>
    <Include file="Networks.xml"/>
    <Include file="Simulation.xml"/>

    <Include file="izhikevich2007_network.nml"/>

    <Simulation id="example_izhikevich2007network_sim" length="1000ms" step="0.1ms" target="IzNet" seed="123">  <!-- Note seed: ensures same random numbers used every run -->
        <EventOutputFile id="Spikes_file__IzhPop0" fileName="{spike_file_name}" format="ID_TIME">
            <EventSelection id="0" select="IzhPop0[0]" eventPort="spike">
            </EventSelection>
            <EventSelection id="1" select="IzhPop0[1]" eventPort="spike">
            </EventSelection>
            <EventSelection id="2" select="IzhPop0[2]" eventPort="spike">
            </EventSelection>
            <EventSelection id="3" select="IzhPop0[3]" eventPort="spike">
            </EventSelection>
            <EventSelection id="4" select="IzhPop0[4]" eventPort="spike">
            </EventSelection>
            <EventSelection id="5" select="IzhPop0[5]" eventPort="spike">
            </EventSelection>
        </EventOutputFile>
        <EventOutputFile id="Spikes_file__IzhPop1" fileName="{spike_file_name}" format="ID_TIME">
            <EventSelection id="6" select="IzhPop1[0]" eventPort="spike">
            </EventSelection>
            <EventSelection id="7" select="IzhPop1[1]" eventPort="spike">
            </EventSelection>
            <EventSelection id="8" select="IzhPop1[2]" eventPort="spike">
            </EventSelection>
            <EventSelection id="9" select="IzhPop1[3]" eventPort="spike">
            </EventSelection>
            <EventSelection id="10" select="IzhPop1[4]" eventPort="spike">
            </EventSelection>
        </EventOutputFile>
        
        

    </Simulation>

</Lems>
    """ 
    this was the lems file that I used

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -381,13 +328,214 @@ def run(a=None, **kwargs):
+            times = []
+            ids = []
+
+            for line in spikes_file:
```
I wonder if this can be made more efficient/shorter using something like `numpy.loadtxt`? Load it using that, then check `format` to see if it's `id time` or `time id`, and depending on that go:  `times = numpyarray[:, 0]` etc?

(this is also correct, but using numpy is probably "nicer").

sanjayankur31 commented on tests/plot/test_PlotSpikes.py:
```diff
@@ -0,0 +1,252 @@
+        spike_file.close()
+
+        # Generate a plot from the spike time file and save it to a file
+        pyplts.plot_spikes(
```
This bit needs fixing---it's not reading from the file?

sanjayankur31 commented on tests/plot/test_PlotSpikes.py:
```diff
@@ -0,0 +1,252 @@
+            base_dir=".",
+            show_plots_already=False,
+            save_spike_plot_to="spikes-test-from-lems-file-2.png",
+            rates=True,
```
we'll have to look at the implementation of `rates` later, not sure it's doing the right thing. Don't worry about it for this PR, though.

sanjayankur31 reviewed (changes requested):
Getting there. A few things to check/tweak (see the inline comments). 

Here are two spike files from the [example in the documentation](https://docs.neuroml.org/Userdocs/IzhikevichNetworkExample.html)

[example_izhikevich2007network_sim.0.spikes.dat.txt](https://github.com/NeuroML/pyNeuroML/files/15236156/example_izhikevich2007network_sim.0.spikes.dat.txt)
[example_izhikevich2007network_sim.1.spikes.dat.txt](https://github.com/NeuroML/pyNeuroML/files/15236157/example_izhikevich2007network_sim.1.spikes.dat.txt)

When plotted together, the spikes overlap because in each population, the ids start from 0. 
![image](https://github.com/NeuroML/pyNeuroML/assets/102575/b97875d1-de7d-40c6-85ae-c176537de53e)


So, I think similar to the plot_time_series function, we need to add an option to add an "offset" that will move each new set of spikes by the amount necessary to prevent overlapping.

AdityaBITMESRA commented on pyneuroml/lems/__init__.py:
```diff
@@ -631,17 +636,11 @@ def load_sim_data_from_lems_file(
-            raise ValueError("No events found")
-        return traces, events
-    elif get_events:
+    if get_events:
```
yes sir I think it was the raising the ValueError No traces found...but I will go through it again to make it more correct.

AdityaBITMESRA replied:
this sounds correct sir, I will make the changes

sanjayankur31 commented:
We do need to infer the total number of cells in each output file

- when plotting output files, we can only infer based on the max cell id
- when plotting from lems files, we can read the model spec and get the total number of cells

sanjayankur31 commented:
Hi @AdityaBITMESRA , please drop a comment here when this is ready for another round of review. 

Oh, and please pull in `development` to your branch to ensure you have the latest code from the other parts of the repo.

AdityaBITMESRA commented:
Sure sir @sanjayankur31 I will keep this in my mind

AdityaBITMESRA commented:
Sir @sanjayankur31 I have addressed all the comments. Can you give it a look?

sanjayankur31 reviewed (commented):
Looking good, but I think we need to test it a little more with real simulation data to be sure. In real simulations, for each population, the cell id will start with 0, so pop1 will be from 0-N, then pop2 will be from 0-M. In this case the current code doesn't quite work:

![image](https://github.com/NeuroML/pyNeuroML/assets/102575/d039e1c5-a8c2-48ab-a773-cf7a26cefe38)

The spikes overlap here ^

These are the two spike files that were used:

[I_net.spikes.txt](https://github.com/NeuroML/pyNeuroML/files/15391858/I_net.spikes.txt)
[E_net.spikes.txt](https://github.com/NeuroML/pyNeuroML/files/15391859/E_net.spikes.txt)

```
pynml-plotspikes -format t_id E_net.spikes.txt I_net.spikes.txt
```

This is data from this model:

https://github.com/OpenSourceBrain/Brunel2000

You can generate it yourself also:

- clone the model
- go into the NeuroML2 folder
- simulate the model with `pynml`: `pynml -nogui LEMS_Sim_BrunelFromPyNN.xml`

I've opened a PR here to pull the latest `development` branch into your PR branch:

https://github.com/AdityaBITMESRA/pyNeuroML/pull/1

Your PR branch still says pyneuroml version 1.2.10, but development is on 1.2.13 now. This PR will update your branch with the new changes. For future reference, here's how you would do it yourself:

```
git remote add upstream https://github.com/NeuroML/pyNeuroML.git  # add main repo as new remote
git fetch upstream # fetch upstream repo state
git checkout development # check out development branch
git pull upstream development # pull upstream remote's development into your local development banch
git push development # push to your fork
git checkout feat/issue-303  # checkout your branch
git merge development  # pull development's changes into it
```

sanjayankur31 commented on pyneuroml/plot/PlotSpikes.py:
```diff
@@ -187,139 +191,92 @@ def read_sonata_spikes_hdf5_file(file_name: str):
-
-                # print("max_id_here in %s: %i"%(file_name, max_id_here))
-                labels.append("%s, %s (%i)" % (name, pop, max_id_here - offset_id))
-                offset_id = max_id_here + 1
```
I think you've missed this out in your refactoring---for each population, the offset needs to be updated, no?

AdityaBITMESRA commented:
Thankyou for all the inputs

sanjayankur31 commented:
> Thankyou for all the inputs

No problem, it's getting there, just a few tweaks. Well done so far!

(If anything is unclear, please feel free to ping me in the [chat channel](https://matrix.to/#/#NeuroML_community:gitter.im). I'm usually there all day from 0900 - 1800 London time on work days)

AdityaBITMESRA commented:
okay sir

sanjayankur31 commented:
> okay sir

Cool. Also, no need to refer to any of us as "sir" :)

sanjayankur31 commented:
When you want me to look at this @AdityaBITMESRA , please request another review (click the button next to my user in the review bit), or just drop me a comment here saying so---just so that I know that I should look at the PR again.

sanjayankur31 commented:
Cool, I'll do another round of review tomorrow.
